### ISSUE TITLE
Must be fixed: Failure to manage time-controls! - Games often lost on draw-positions

### ISSUE THREAD

**Issue by bitsocket** [lantonov/asmFish#58]
There must be a problem with management of time-controls - asmFishW (v2017-05-15) repeatedly loses games (while it has more material and position advantage) just because of flag!

The engine plays a drawn position (with about 1 min left on the clock) and tries to win it (without real attempt), until it finally loses on time-controls(!) For example, see this hillarious game (asmFishW plays black pieces):

[Game-lost-on-Draw-Position.pgn.zip](https://github.com/lantonov/asmFish/files/1026782/Game-lost-on-Draw-Position.pgn.zip)

Even the weakest player could not have lost this game, **because Black could (and should) have simply given away its pieces to make it a draw – by chess rules!**

Final position: (Game was played with 5-men Syzygy tables.)
![final position black loses](https://cloud.githubusercontent.com/assets/22038957/26384701/4cfa02ca-403b-11e7-9703-d8c9a490f877.png)
- **Black** (asmFish) managed to lose this position on time-controls.

**Comment by lantonov** [comment]
Since the time control is with no increment (300+0), I think that the other engine beat asmFish on time in a dead drawn position just by postponing draw (no 3-fold rep) and relying on its smaller time allocation per move. I don't see this as a bug in asmFish time management. Making asmFish move faster will reflect negatively on its strength. Making it recognise dead draws and forcing draws after some number of moves will also make it weaker because it may miss wins. 
Giving up pawns may not be effective if the opponent wants to avoid the draw. For instance, in the above position, the opponent may take the first pawn just to avoid the 50-move rule and delay taking the second pawn as long as possible.

**Comment by bitsocket** [comment]
Sorry, I disagree completely. Black can never ever win this position but must manage not to lose on time.
It is not about moving faster. It's about recognizing that it's dead draw, despite material advantage. And I am sure the engine did recognize it's a dead draw position - with or also without 5-men SYZYGY. 

There is no possibility for White to postpone taking Black's pawns once they go forward. And the thing here is not that the engine disrespected the position and "forgot" time-controls, _during just a series of moves within a few seconds_. No, the engine had plenty of time and it several times took much longer to calculate the next (stupid) move than the opponent. The  whole drama went on for about 20 moves during 1 minute thinking-time for Black. Just replay the game from the PGN and judge for yourself. 

I understand it will be probably difficult to fix this. But there is no point in saying this would not be a bug (but a feature). The minimum should be, that the engine at least offers a draw in such positions - after a series of moves without gaining any improvement towards a win - over a certain period of time.

BTW, I never noticed such losses like this with Stockfish. And I am sure Stockfish would not have "tried to win" without having moved it's pawns. If Black wanted to win the game - **and not just wait for flag** - he should have at least tried it.

I don't know whether you will take the effort to analyze the problem. But what I do know is, that this bug is a serious weakness with respect to asmFish's playing-strength.

Best regards.

**Comment by lantonov** [comment]
Ok, I will analyze this game with asmFish and Stockfish in parallel and see what is going on.

**Comment by bitsocket** [comment]
Ok, great, I appreciate that.

_I see it as follows:_
The black King must aim to reach f2 (via e3). White can only avoid this by moving it's King to d3, but than the c5 pawn can progress to c4 (on suicide mission). Finally, the white bishop must beat the black pawn on g2, while the black King got to f2 - and the **draw is enforced** (without waiting for 50 move rule).

**Comment by tthsqe12** [comment]
It must be realized that in such positions stockfish is going to reach very high depths (im guessing around 60, or even the max 127). In this case it prints a pv line for each depth, and if your gui is even the least bit sluggish in processing this output, time losses are going to be inevitable without increment. What we can do is make a more aggressive spam filter and set USE_SPAMFILTER = 1 as the default.

**Comment by lantonov** [comment]
Starting from the position just before move 139 (after the rooks disappear), I pitched Stockfish with white against asmFish with black in Winboard. Time control: 60 sec + 0 sec and 5-men syzygy TB. Draw by insufficient material. Before the draw Stockfish had 3.2 sec left and asmFish had 6.1 sec. The observed loss couldn't be reproduced. Most of the time both engines thought that black is winning with almost equal evaluations (1.08-1.09). Below is the game:
```
[Event "Computer Chess Game"]
[Site "DESKTOP-RL8DOC5"]
[Date "2017.05.25"]
[Round "-"]
[White "Stockfish 200517 64 BMI2"]
[Black "asmFishW_2017-05-22_bmi2"]
[Result "1/2-1/2"]
[TimeControl "60"]
[FEN "8/8/6k1/1Kp5/1b6/5Bp1/8/8 w - - 0 1"]
[SetUp "1"]

{--------------
. . . . . . . .
. . . . . . . .
. . . . . . k .
. K p . . . . .
. b . . . . . .
. . . . . B p .
. . . . . . . .
. . . . . . . .
white to play
--------------}
1. Kc4 {-1.09/69} Kg5 {+1.08/59 1.3} 2. Kd3 {-1.09/59 0.9} Kf4 {+1.08/63}
3. Ba8 {-1.09/61 0.9} Be1 {+1.08/61 0.7} 4. Bh1 {-1.09/59 1.0} Ba5
{+1.08/64 0.7} 5. Bg2 {-1.09/62 1.0} Bc7 {+1.08/65} 6. Bd5 {-1.09/62 0.9}
Bb6 {+1.08/65 0.7} 7. Bc6 {-1.09/69 1.0} Kg5 {+1.08/64 0.7} 8. Kc3
{-1.09/64 1.0} Kg4 {+1.08/64 0.7} 9. Bg2 {-1.09/68 1.0} Bc7 {+1.08/71} 10.
Kd3 {-1.09/66 0.9} Kf4 {+1.08/70 0.8} 11. Kc4 {-1.09/63 1.0} Bb6
{+1.08/71 6} 12. Kd3 {-1.09/71 1.0} Ba7 {+1.08/71 4} 13. Bh1 {-1.09/67 0.9}
Ke5 {+1.08/69 0.6} 14. Bg2 {-1.09/70 1.2} Kf5 {+1.08/66} 15. Bf1
{-1.09/67 0.9} Ke5 {+1.08/63 0.6} 16. Bg2 {-1.09/61 1.0} Bb6 {+1.08/70} 17.
Kc4 {-1.09/64 2.4} Kf5 {+1.08/65 2.9} 18. Bc6 {-1.09/64 2.3} Kg4
{+1.08/58 2.8} 19. Bg2 {-1.09/61 7} Kf4 {+1.08/58} 20. Kd3 {-1.09/56 0.1}
Bd8 {+1.08/56} 21. Bd5 {-1.09/55 6} Be7 {+1.08/57 0.8} 22. Bh1
{-1.09/53 1.6} Ke5 {+1.08/53 5} 23. Bg2 {-1.09/54 5} Bf8 {+1.08/51 4} 24.
Bh1 {-1.09/52} Be7 {+1.08/49 4} 25. Bg2 {-1.09/51 4} Kf5 {+1.08/47} 26. Kc4
{-1.09/48 3} Kg4 {+1.08/46 3} 27. Kc3 {-1.09/46 0.8} Kf4 {+1.08/47 3} 28.
Kd3 {-1.09/46 0.6} Bd6 {+1.08/45} 29. Bc6 {-1.09/41 2.2} Bc7 {+1.08/43 2.6}
30. Kc4 {-1.09/38 1.7} Bb6 {+1.08/39 2.2} 31. Kd3 {-1.02/39 1.4} Kg5
{+1.08/41} 32. Bg2 {-1.09/36 1.0} Kf5 {+1.00/36 0.8} 33. Kc3 {-1.02/35 0.8}
Bd8 {+1.00/32 1.7} 34. Kc4 {-1.02/32 0.1} Be7 {+1.00/30 1.3} 35. Kd3
{-1.09/32} Kg5 {+1.08/31 1.2} 36. Bd5 {-1.09/30 0.6} Kf4 {+1.00/29 0.4} 37.
Bg2 {-1.00/26 0.2} Bg5 {+0.98/27 0.8} 38. Kc4 {+0.00/61 0.1} Bh4
{+0.00/86 0.2} 39. Kxc5 {+0.00/74 0.1} Ke3 {+0.00/14} 40. Kc4
{+0.00/14 0.1} Kf2 {+0.00/21 0.2} 41. Bd5 {+0.00/15 0.2} Kf1 {+0.00/26 0.1}
42. Kd4 {+0.00/15} Bd8 {+0.00/26 0.4} 43. Ke3 {+0.00/22 0.1} Kg1
{+0.00/24 0.1} 44. Kf4 {+0.00/20 0.1} Kf2 {+0.00/24 0.2} 45. Kf5
{+0.00/23 0.1} Bc7 {+0.00/27 0.1} 46. Kg4 {+0.00/23 0.1} Bd6 {+0.00/29 0.1}
47. Bb7 {+0.00/25 0.1} Be5 {+0.00/27 0.1} 48. Kh3 {+0.00/28 0.1} Bb2
{+0.00/30} 49. Ba8 {+0.00/29} Bd4 {+0.00/36 0.2} 50. Bh1 {+0.00/28} Bf6
{+0.00/33 0.1} 51. Bg2 {+0.00/29 0.1} Ba1 {+0.00/39 0.1} 52. Ba8 {+0.00/25}
Bd4 {+0.00/38 0.1} 53. Kh4 {+0.00/25 0.1} g2 {+0.00/20 0.1} 54. Bxg2
{+0.00/2 0.1} Kxg2 {+0.00/26}
{Xboard adjudication: Insufficient mating material} 1/2-1/2
```

**Comment by bitsocket** [comment]
In the re-stage of the game ([Game-End-Replay.pgn.zip](https://github.com/lantonov/asmFish/files/1030162/Game-End-Replay.pgn.zip)) asmFish moved his g-pawn (and forced the draw) **while still having 6+ seconds** left on the clock - very good. It seems the program can handle it right - sometimes.

I just wonder then, why in the actual game the engine went on with useless moves until the very last split-second and then finally lost it on time?

Maybe it's just Winboard being more responsive than Fritz-Gui and the time-lag had accumulated.
I guess this is what tthsqe12 is referring to when he mentions the "pv-line" and adjusting the "SPAM_FILTER" (short before time runs out?).

**Comment by lantonov** [comment]
Stockfish and asmFish have the habit to spam the GUI especially towards the endgame when they send long PVs which the GUI doesn't have enough time to process through the UCI. The GUI then takes the lost time from the engine thinking time and then it is possible for the engine to lose on time. I guess that in the actual game asmFish didn't have enough time towards the end to realise that the evaluation is 0.00 and it was still thinking that it was 1.08 so it was reluctant to give its pawns.
